easybuild/tools/build_from_specsfile.py

```diff
@@ -0,0 +1,150 @@
+        easyconfigs_full_paths = []
+        for sw in self.software_list:
+            path_to_append = self.get_ec_path(sw)
+
+            if path_to_append is None:
+                continue
+            else:
+                easyconfigs_full_paths.append(path_to_append)
+        return easyconfigs_full_paths
+
+    # single command
+    def get_ec_path(self, sw):
+        full_path = search_easyconfigs(query='%s-%s-%s-%s' % (
+            str(sw.software), str(sw.version), str(sw.toolchain_name), str(sw.toolchain_version)),
+            short=False, filename_only=False,
+            terse=False, consider_extra_paths=True, print_result=False, case_sensitive=False)
+        if len(full_path) == 1:
+            return full_path[0]
+        else:
+            print('%s does not have clearly specified parameters - %s matches found. Skipping. \n' %
+                  (sw.software, str(len(full_path))))
+
+    # todo: flags applicable to all sw (i.e. robot)
+    def get_general_options(self):
+        general_options = {}
+        general_options['robot'] = self.robot
+        general_options['easybuild_version'] = self.easybuild_version
+        return general_options
+
+
+# single sw command
+class SoftwareSpecs(object):
+    def __init__(self, software, version, toolchain, toolchain_version, toolchain_name):
+        self.software = software
+        self.version = version
+        self.toolchain = toolchain
+        self.toolchain_version = toolchain_version
+        self.toolchain_name = toolchain_name
+        self.toolchain = toolchain
+
+        self.version_suffix = None
+
+    def get_version_suffix(self):
+        return self.version_suffix or ''
+
+
+# implement this to your own needs - to create custom yaml/json/xml parser
+class GenericSpecsParser(object):
+    @staticmethod
+    def parse(filename):
+        raise NotImplementedError
+
+
+class YamlSpecParser(GenericSpecsParser):
+    @staticmethod
+    def parse(filename):
+
+        try:
+            with open(filename, 'r') as f:
+                spec_dict = yaml.safe_load(f)
+
+            eb = Specsfile()
+        except FileNotFoundError:
+            raise EasyBuildError("Could not read provided specsfile.")
+
+        sw_dict = spec_dict["software"]
+
+        # assign software-specific EB attributes
+        for software in sw_dict:
+            try:
+                # iterates through toolchains to find out what sw version is needed
+                for yaml_toolchain in sw_dict[software]['toolchains']:
+                    # retrieves version number
+                    for yaml_version in sw_dict[software]['toolchains'][yaml_toolchain]['versions']:
+
+                        # if among versions there is anything else than known flags or version flag itself, it is wrong
+                        # identification of version strings is vague
+                        if str(yaml_version)[0].isdigit() \
+                                or str(yaml_version)[-1].isdigit():
```

Comment:
Is dangerous to make a lot of assumptions here, you wouldn't believe the things we've seen... :)

Even assuming that a software version starts or ends with a digit can lead to limitations we don't want.

I would just handle the special cases (like `versionsuffix` and `exclude-labels`/`include-labels`) first, and if there are not other matches assume you've found a software version, without checking the value at all...

So:

```python
if key == 'versionsuffix':
    ...
elif key in ('exclude-labels', 'include-labels'):
    ...
else:
    # assume the value is a software version
```

In the last case you can add a check to ensure the software version wasn't found yet.
If you hit two keys that would be interpreted as a software version, something is wrong?